numpy/core/src/multiarray/ctors.c

```diff
@@ -1183,28 +1183,43 @@ PyArray_NewFromDescrAndBase(
                                     flags, obj, base, 0, 0);
 }
 
-/*NUMPY_API
+/*
  * Creates a new array with the same shape as the provided one,
- * with possible memory layout order and data type changes.
+ * with possible memory layout order, data type and shape changes.
  *
  * prototype - The array the new one should be like.
  * order     - NPY_CORDER - C-contiguous result.
  *             NPY_FORTRANORDER - Fortran-contiguous result.
  *             NPY_ANYORDER - Fortran if prototype is Fortran, C otherwise.
  *             NPY_KEEPORDER - Keeps the axis ordering of prototype.
  * dtype     - If not NULL, overrides the data type of the result.
+ * ndim      - If not 0 and dims not NULL, overrides the shape of the result.
+ * dims      - If not NULL and ndim not 0, overrides the shape of the result.
  * subok     - If 1, use the prototype's array subtype, otherwise
  *             always create a base-class array.
  *
  * NOTE: If dtype is not NULL, steals the dtype reference.  On failure or when
  * dtype->subarray is true, dtype will be decrefed.
  */
 NPY_NO_EXPORT PyObject *
-PyArray_NewLikeArray(PyArrayObject *prototype, NPY_ORDER order,
-                     PyArray_Descr *dtype, int subok)
+PyArray_NewLikeArrayWithShape(PyArrayObject *prototype, NPY_ORDER order,
+                              PyArray_Descr *dtype, int ndim, npy_intp *dims, int subok)
 {
     PyObject *ret = NULL;
-    int ndim = PyArray_NDIM(prototype);
+
+    int new_shape = 1;
+
+    if (dims == NULL) {
+        ndim = PyArray_NDIM(prototype);
+        dims = PyArray_DIMS(prototype);
+        new_shape = 0;
+    }
+    else if (order == NPY_KEEPORDER && (ndim != PyArray_NDIM(prototype) ||
```

Comment:
This breaks previous behavior which did not require the prototype to be contiguous.
```
In [1]: a = ones((3,4))[:,::2]

In [2]: a.flags
Out[2]: 
  C_CONTIGUOUS : False
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False

In [3]: ones_like(a, order='K')
Out[3]: 
array([[1., 1.],
       [1., 1.],
       [1., 1.]])
```
I'm not sure why an error should be raised, why not just override the shape and go with C order when a new shape is specified?

Reply:
@charris what you're saying is like my original implementation. I modified this based on @eric-wieser's request here https://github.com/numpy/numpy/pull/13046#discussion_r260673995. For me it's fine either way.

Reply:
@charris: This shouldn't break any existing code (assuming my comment below is addressed).

The thinking is that requesting the order (of the strides) to be kept is nonsensical if the the number of strides is not the same.